src/main/java/teammates/logic/api/EmailGenerator.java

```diff
@@ -1,2 +1,1083 @@
+        List<InstructorAttributes> instructorsToNotify = isEmailNeededForStudents
+                ? instructorsLogic.getCoOwnersForCourse(session.getCourseId())
+                : new ArrayList<>();
+        List<StudentAttributes> students = isEmailNeededForStudents
+                ? studentsLogic.getStudentsForCourse(session.getCourseId())
+                : new ArrayList<>();
+        List<InstructorAttributes> instructors = isEmailNeededForInstructors
+                ? instructorsLogic.getInstructorsForCourse(session.getCourseId())
+                : new ArrayList<>();
+
+        if (emailType == EmailType.FEEDBACK_CLOSING) {
+            Map<String, Instant> studentDeadlines = session.getStudentDeadlines();
+            students = students.stream()
+                    .filter(x -> !studentDeadlines.containsKey(x.getEmail()))
+                    .collect(Collectors.toList());
+
+            Map<String, Instant> instructorDeadlines = session.getInstructorDeadlines();
+            instructors = instructors.stream()
+                    .filter(x -> !instructorDeadlines.containsKey(x.getEmail()))
+                    .collect(Collectors.toList());
+        }
+
+        String status = emailType == EmailType.FEEDBACK_OPENING
+                ? FEEDBACK_STATUS_SESSION_OPENING
+                : FEEDBACK_STATUS_SESSION_CLOSING;
+
+        String template = emailType == EmailType.FEEDBACK_OPENING
+                ? EmailTemplates.USER_FEEDBACK_SESSION_OPENING.replace("${status}", status)
+                : EmailTemplates.USER_FEEDBACK_SESSION.replace("${status}", status);
+
+        return generateFeedbackSessionEmailBases(course, session, students, instructors, instructorsToNotify, template,
+                emailType, FEEDBACK_ACTION_SUBMIT_EDIT_OR_VIEW);
+    }
+
+    /**
+     * Generates the feedback session opening soon emails for the given {@code session}.
+     *
+     * <p>This is useful for e.g. in case the feedback session opening info was set wrongly.
+     */
+    public List<EmailWrapper> generateFeedbackSessionOpeningSoonEmails(FeedbackSessionAttributes session) {
+        return generateFeedbackSessionOpeningSoonOrClosedEmails(session, EmailType.FEEDBACK_OPENING_SOON);
+    }
+
+    private List<EmailWrapper> generateFeedbackSessionOpeningSoonOrClosedEmails(
+            FeedbackSessionAttributes session, EmailType emailType) {
+        CourseAttributes course = coursesLogic.getCourse(session.getCourseId());
+        // Notify only course co-owners
+        List<InstructorAttributes> coOwners = instructorsLogic.getCoOwnersForCourse(session.getCourseId());
+        return coOwners.stream()
+                .map(coOwner -> generateFeedbackSessionEmailBaseForCoowner(course, session, coOwner, emailType))
+                .collect(Collectors.toList());
+    }
+
+    private EmailWrapper generateFeedbackSessionEmailBaseForCoowner(
```

Comment:
This method is very long and has a lot of conditional logic, consider dividing it up into a few different methods to reduce length and conditionals.